internal/controller/endpoint/controller_finalize.go

```diff
@@ -65,12 +68,29 @@ func (r *Reconciler) finalize(ctx context.Context, old, ep *choreov1.Endpoint) (
 	}
 
 	resourceHandlers := r.makeExternalResourceHandlers()
+	pendingDeletion := false
+
 	for _, resourceHandler := range resourceHandlers {
-		if err := resourceHandler.Delete(ctx, epCtx); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to delete external resource %s: %w", resourceHandler.Name(), err)
+		exists, err := resourceHandler.GetCurrentState(ctx, epCtx)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to get current state of external resource %s: %w", resourceHandler.Name(), err)
+		}
+
+		if exists != nil {
+			pendingDeletion = true
+			// Trigger deletion of the resource as it is still exists
+			if err := resourceHandler.Delete(ctx, epCtx); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to delete external resource %s: %w", resourceHandler.Name(), err)
+			}
 		}
 	}
 
+	// Requeue the reconcile loop if there are still resources pending deletion
+	if pendingDeletion {
+		logger.Info("endpoint deletion is still pending as the dependent resource deletion pending.. retrying..")
```

Comment:
I think this could be an event for the deployment object. But since we doing this with 5-second retries, both logs and events can accumulate a lot with many deployments. Shall we use an exponential delay with only retry=true and see how it goes? We can change it later if required.

Reply:
For the moment, let's avoid events. Since it is an exponential backoff, it will fire many events in the beginning.